File: src/com/leozhi/topic42/Solution03.java
package com.leozhi.topic42;

/**
 * @author leozhi
 * 左右指针
 * 通过
 * 1ms
 */
public class Solution03 {
    public int trap(int[] height) {
        int n = height.length;
        if (n == 0) {
            return 0;
        }
        int left = 0, right = n - 1, res = 0;
        int leftMax = height[0], rightMax = height[n - 1];
        while (left <= right) {
            leftMax = Math.max(leftMax, height[left]);
            rightMax = Math.max(rightMax, height[right]);
            if (leftMax < rightMax) {
                res += leftMax - height[left];
                left++;
            } else {
                res += rightMax - height[right];
                right--;
            }
        }
        return res;
    }
}

File: src/com/leozhi/topic633/Solution03.kt
package com.leozhi.topic633

import kotlin.math.sqrt

/**
 * @author leozhi
 * 左右指针
 * 通过
 * 188ms
 */
class Solution03 {
    fun judgeSquareSum(c: Int): Boolean {
        val len = sqrt(c.toDouble()).toInt() + 1
        val pow = IntArray(len)
        for (i in 0 until len) {
            pow[i] = i * i
        }
        var left = 0
        var right = len - 1
        while (left <= right) {
            when {
                pow[left] + pow[right] == c -> {
                    return true
                }
                pow[left] + pow[right] < c -> {
                    ++left
                }
                else -> {
                    --right
                }
            }
        }
        return false
    }
}